Eval bug: Kill switch tripped while encoding audio slices (server-context.cpp:2686: fatal error - please provide logs)

### Name and Version

% ./llama-server --version
ggml_metal_device_init: tensor API disabled for pre-M5 and pre-A19 devices
ggml_metal_library_init: using embedded metal library
ggml_metal_library_init: loaded in 0.008 sec
ggml_metal_rsets_init: creating a residency set collection (keep_alive = 180 s)
ggml_metal_device_init: GPU name:   MTL0
ggml_metal_device_init: GPU family: MTLGPUFamilyApple9  (1009)
ggml_metal_device_init: GPU family: MTLGPUFamilyCommon3 (3003)
ggml_metal_device_init: GPU family: MTLGPUFamilyMetal4  (5002)
ggml_metal_device_init: simdgroup reduction   = true
ggml_metal_device_init: simdgroup matrix mul. = true
ggml_metal_device_init: has unified memory    = true
ggml_metal_device_init: has bfloat            = true
ggml_metal_device_init: has tensor            = false
ggml_metal_device_init: use residency sets    = true
ggml_metal_device_init: use shared buffers    = true
ggml_metal_device_init: recommendedMaxWorkingSetSize  = 55662.79 MB
version: 8562 (c46758d28)
built with AppleClang 17.0.0.17000604 for Darwin arm64

### Operating systems

Mac

### GGML backends

Metal

### Hardware

Apple Mac Mini M4 Pro 64GB

### Models

MistralAI Voxtral Mini 3B 2507 Q8

https://huggingface.co/bartowski/mistralai_Voxtral-Mini-3B-2507-GGUF


### Problem description & steps to reproduce

The cause of this error is the new code added in `107d599`, it looks like there were issues where the server would get stuck, but in this case the server is busy encoding the audio slices, and after 4 "ticks" of encoding, the kill switch is tripped.

To reproduce, start the server:

```bash
./llama-server
  --model /llm/mistralai_Voxtral-Mini-3B-2507/mistralai_Voxtral-Mini-3B-2507-Q8_0.gguf
  --mmproj /llm/mistralai_Voxtral-Mini-3B-2507/mmproj-mistralai_Voxtral-Mini-3B-2507-f16.gguf
  --ctx-size 16384
  --temp 0.0
  --top-p 0.95
```

Then use the webui to upload an mp3 along with a prompt: `summarize the transcript`. The encoding starts as you can see in the console, but after 4 ticks the kill switch is tripped and the fatal error occurs.

### First Bad Commit

% git bisect bad
107d5999520dd02195ebe05278752db9fd33c865 is the first bad commit
commit 107d5999520dd02195ebe05278752db9fd33c865 (HEAD, tag: b8254)
Author: Georgi Gerganov <ggerganov@gmail.com>
Date:   Mon Mar 9 10:33:12 2026 +0200

    server : add kill switch when server is stuck (#20277)

 tools/server/server-context.cpp | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

As this was my first time using `git bisect`, to confirm I checked out the previous version with `git checkout 107d5999520dd02195ebe05278752db9fd33c865^`, built and confirmed that `8253 (e8bbc736c)` is the last good version and 8254 (107d59995) is the first bad one.

### Relevant log output

<details>
<summary>Logs</summary>
<!-- Copy-pasted short logs go into the "console" area here -->

```console
ggml_metal_device_init: tensor API disabled for pre-M5 and pre-A19 devices
ggml_metal_library_init: using embedded metal library
ggml_metal_library_init: loaded in 0.005 sec
ggml_metal_rsets_init: creating a residency set collection (keep_alive = 180 s)
ggml_metal_device_init: GPU name:   MTL0
ggml_metal_device_init: GPU family: MTLGPUFamilyApple9  (1009)
ggml_metal_device_init: GPU family: MTLGPUFamilyCommon3 (3003)
ggml_metal_device_init: GPU family: MTLGPUFamilyMetal4  (5002)
ggml_metal_device_init: simdgroup reduction   = true
ggml_metal_device_init: simdgroup matrix mul. = true
ggml_metal_device_init: has unified memory    = true
ggml_metal_device_init: has bfloat            = true
ggml_metal_device_init: has tensor            = false
ggml_metal_device_init: use residency sets    = true
ggml_metal_device_init: use shared buffers    = true
ggml_metal_device_init: recommendedMaxWorkingSetSize  = 55662.79 MB
main: n_parallel is set to auto, using n_parallel = 4 and kv_unified = true
build: 8562 (c46758d28) with AppleClang 17.0.0.17000604 for Darwin arm64
system info: n_threads = 10, n_threads_batch = 10, total_threads = 14

system_info: n_threads = 10 (n_threads_batch = 10) / 14 | MTL : EMBED_LIBRARY = 1 | CPU : NEON = 1 | ARM_FMA = 1 | FP16_VA = 1 | MATMUL_INT8 = 1 | DOTPROD = 1 | SME = 1 | ACCELERATE = 1 | REPACK = 1 | 

Running without SSL
init: using 13 threads for HTTP server
start: binding port with default address family
main: loading model
srv    load_model: loading model '/llm/mistralai_Voxtral-Mini-3B-2507/mistralai_Voxtral-Mini-3B-2507-Q8_0.gguf'
common_init_result: fitting params to device memory, for bugs during this step try to reproduce them with -fit off, or provide --verbose logs if the bug only occurs with -fit on
llama_params_fit_impl: projected to use 5847 MiB of device memory vs. 53083 MiB of free device memory
llama_params_fit_impl: will leave 47235 >= 1024 MiB of free device memory, no changes needed
llama_params_fit: successfully fit params to free device memory
llama_params_fit: fitting params to free memory took 0.13 seconds
llama_model_load_from_file_impl: using device MTL0 (Apple M4 Pro) (unknown id) - 53083 MiB free
llama_model_loader: loaded meta data with 40 key-value pairs and 273 tensors from /llm/mistralai_Voxtral-Mini-3B-2507/mistralai_Voxtral-Mini-3B-2507-Q8_0.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Voxtral Mini 3B 2507
llama_model_loader: - kv   3:                            general.version str              = 2507
llama_model_loader: - kv   4:                           general.basename str              = Voxtral-Mini
llama_model_loader: - kv   5:                         general.size_label str              = 3B
llama_model_loader: - kv   6:                            general.license str              = apache-2.0
llama_model_loader: - kv   7:                               general.tags arr[str,1]       = ["vllm"]
llama_model_loader: - kv   8:                          general.languages arr[str,8]       = ["en", "fr", "de", "es", "it", "pt", ...
llama_model_loader: - kv   9:                          llama.block_count u32              = 30
llama_model_loader: - kv  10:                       llama.context_length u32              = 131072
llama_model_loader: - kv  11:                     llama.embedding_length u32              = 3072
llama_model_loader: - kv  12:                  llama.feed_forward_length u32              = 8192
llama_model_loader: - kv  13:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv  14:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv  15:                       llama.rope.freq_base f32              = 100000000.000000
llama_model_loader: - kv  16:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  17:                 llama.attention.key_length u32              = 128
llama_model_loader: - kv  18:               llama.attention.value_length u32              = 128
llama_model_loader: - kv  19:                           llama.vocab_size u32              = 131072
llama_model_loader: - kv  20:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv  21:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  22:                         tokenizer.ggml.pre str              = tekken
llama_model_loader: - kv  23:                      tokenizer.ggml.merges arr[str,269443]  = ["ƒ† ƒ†", "ƒ† t", "e r", "i n", "ƒ† ƒ...
llama_model_loader: - kv  24:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  25:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  26:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  27:            tokenizer.ggml.padding_token_id u32              = 11
llama_model_loader: - kv  28:                      tokenizer.ggml.tokens arr[str,131072]  = ["<unk>", "<s>", "</s>", "[INST]", "[...
llama_model_loader: - kv  29:                      tokenizer.ggml.scores arr[i32,131072]  = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
llama_model_loader: - kv  30:                  tokenizer.ggml.token_type arr[i32,131072]  = [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, ...
llama_model_loader: - kv  31:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  32:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  33:                    tokenizer.chat_template str              = {#- Copyright 2025-present the Unslot...
llama_model_loader: - kv  34:               general.quantization_version u32              = 2
llama_model_loader: - kv  35:                          general.file_type u32              = 7
llama_model_loader: - kv  36:                      quantize.imatrix.file str              = /models_out/Voxtral-Mini-3B-2507-GGUF...
llama_model_loader: - kv  37:                   quantize.imatrix.dataset str              = /training_dir/calibration_datav3.txt
llama_model_loader: - kv  38:             quantize.imatrix.entries_count u32              = 210
llama_model_loader: - kv  39:              quantize.imatrix.chunks_count u32              = 499
llama_model_loader: - type  f32:   61 tensors
llama_model_loader: - type q8_0:  212 tensors
print_info: file format = GGUF V3 (latest)
print_info: file type   = Q8_0
print_info: file size   = 3.97 GiB (8.50 BPW) 
load: 0 unused tokens
load: printing all EOG tokens:
load:   - 2 ('</s>')
load: special tokens cache size = 1000
load: token to piece cache size = 0.8498 MB
print_info: arch                  = llama
print_info: vocab_only            = 0
print_info: no_alloc              = 0
print_info: n_ctx_train           = 131072
print_info: n_embd                = 3072
print_info: n_embd_inp            = 3072
print_info: n_layer               = 30
print_info: n_head                = 32
print_info: n_head_kv             = 8
print_info: n_rot                 = 128
print_info: n_swa                 = 0
print_info: is_swa_any            = 0
print_info: n_embd_head_k         = 128
print_info: n_embd_head_v         = 128
print_info: n_gqa                 = 4
print_info: n_embd_k_gqa          = 1024
print_info: n_embd_v_gqa          = 1024
print_info: f_norm_eps            = 0.0e+00
print_info: f_norm_rms_eps        = 1.0e-05
print_info: f_clamp_kqv           = 0.0e+00
print_info: f_max_alibi_bias      = 0.0e+00
print_info: f_logit_scale         = 0.0e+00
print_info: f_attn_scale          = 0.0e+00
print_info: n_ff                  = 8192
print_info: n_expert              = 0
print_info: n_expert_used         = 0
print_info: n_expert_groups       = 0
print_info: n_group_used          = 0
print_info: causal attn           = 1
print_info: pooling type          = -1
print_info: rope type             = 0
print_info: rope scaling          = linear
print_info: freq_base_train       = 100000000.0
print_info: freq_scale_train      = 1
print_info: n_ctx_orig_yarn       = 131072
print_info: rope_yarn_log_mul     = 0.0000
print_info: rope_finetuned        = unknown
print_info: model type            = 256M
print_info: model params          = 4.01 B
print_info: general.name          = Voxtral Mini 3B 2507
print_info: vocab type            = BPE
print_info: n_vocab               = 131072
print_info: n_merges              = 269443
print_info: BOS token             = 1 '<s>'
print_info: EOS token             = 2 '</s>'
print_info: UNK token             = 0 '<unk>'
print_info: PAD token             = 11 '<pad>'
print_info: LF token              = 1010 'ƒä'
print_info: EOG token             = 2 '</s>'
print_info: max token length      = 150
load_tensors: loading model tensors, this can take a while... (mmap = true, direct_io = false)
load_tensors: offloading output layer to GPU
load_tensors: offloading 29 repeating layers to GPU
load_tensors: offloaded 31/31 layers to GPU
load_tensors:   CPU_Mapped model buffer size =   408.00 MiB
load_tensors:  MTL0_Mapped model buffer size =  4067.96 MiB
..................................................................................
common_init_result: added </s> logit bias = -inf
llama_context: constructing llama_context
llama_context: n_seq_max     = 4
llama_context: n_ctx         = 16384
llama_context: n_ctx_seq     = 16384
llama_context: n_batch       = 2048
llama_context: n_ubatch      = 512
llama_context: causal_attn   = 1
llama_context: flash_attn    = auto
llama_context: kv_unified    = true
llama_context: freq_base     = 100000000.0
llama_context: freq_scale    = 1
llama_context: n_ctx_seq (16384) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M4 Pro
ggml_metal_init: picking default device: Apple M4 Pro
ggml_metal_init: use fusion         = true
ggml_metal_init: use concurrency    = true
ggml_metal_init: use graph optimize = true
llama_context:        CPU  output buffer size =     2.00 MiB
llama_kv_cache:       MTL0 KV buffer size =  1920.00 MiB
llama_kv_cache: size = 1920.00 MiB ( 16384 cells,  30 layers,  4/1 seqs), K (f16):  960.00 MiB, V (f16):  960.00 MiB
sched_reserve: reserving ...
sched_reserve: Flash Attention was auto, set to enabled
sched_reserve: resolving fused Gated Delta Net support:
sched_reserve: fused Gated Delta Net (autoregressive) enabled
sched_reserve: fused Gated Delta Net (chunked) enabled
sched_reserve:       MTL0 compute buffer size =   268.00 MiB
sched_reserve:        CPU compute buffer size =    44.01 MiB
sched_reserve: graph nodes  = 937
sched_reserve: graph splits = 2
sched_reserve: reserve took 7.42 ms, sched copies = 1
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
clip_model_loader: model name:   Voxtral Mini 3B 2507
clip_model_loader: description:  
clip_model_loader: GGUF version: 3
clip_model_loader: alignment:    32
clip_model_loader: n_tensors:    489
clip_model_loader: n_kv:         22

clip_model_loader: has audio encoder
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M4 Pro
ggml_metal_init: picking default device: Apple M4 Pro
ggml_metal_init: use fusion         = true
ggml_metal_init: use concurrency    = true
ggml_metal_init: use graph optimize = true
clip_ctx: CLIP using MTL0 backend
load_hparams: projector:          voxtral
load_hparams: n_embd:             1280
load_hparams: n_head:             20
load_hparams: n_ff:               5120
load_hparams: n_layer:            32
load_hparams: ffn_op:             gelu_erf
load_hparams: projection_dim:     3072

--- audio hparams ---
load_hparams: n_mel_bins:         128
load_hparams: proj_stack_factor:  4
load_hparams: audio_chunk_len:    30
load_hparams: audio_sample_rate:  16000
load_hparams: audio_n_fft:        400
load_hparams: audio_window_len:   400
load_hparams: audio_hop_len:      160

load_hparams: model size:         1267.53 MiB
load_hparams: metadata size:      0.17 MiB
warmup: warmup with audio size = 3000
alloc_compute_meta:       MTL0 compute buffer size =    51.27 MiB
alloc_compute_meta:        CPU compute buffer size =     1.46 MiB
alloc_compute_meta: graph splits = 1, nodes = 992
warmup: flash attention is enabled
init_audio: audio input is in experimental stage and may have reduced quality:
    https://github.com/ggml-org/llama.cpp/discussions/13759
srv    load_model: loaded multimodal model, '/llm/mistralai_Voxtral-Mini-3B-2507/mmproj-mistralai_Voxtral-Mini-3B-2507-f16.gguf'
srv    load_model: initializing slots, n_slots = 4
no implementations specified for speculative decoding
slot   load_model: id  0 | task -1 | speculative decoding context not initialized
slot   load_model: id  0 | task -1 | new slot, n_ctx = 16384
no implementations specified for speculative decoding
slot   load_model: id  1 | task -1 | speculative decoding context not initialized
slot   load_model: id  1 | task -1 | new slot, n_ctx = 16384
no implementations specified for speculative decoding
slot   load_model: id  2 | task -1 | speculative decoding context not initialized
slot   load_model: id  2 | task -1 | new slot, n_ctx = 16384
no implementations specified for speculative decoding
slot   load_model: id  3 | task -1 | speculative decoding context not initialized
slot   load_model: id  3 | task -1 | new slot, n_ctx = 16384
srv    load_model: prompt cache is enabled, size limit: 8192 MiB
srv    load_model: use `--cache-ram 0` to disable the prompt cache
srv    load_model: for more info see https://github.com/ggml-org/llama.cpp/pull/16391
init: chat template, example_format: '[SYSTEM_PROMPT]You are a helpful assistant[/SYSTEM_PROMPT][INST]Hello[/INST]Hi there</s>[INST]How are you?[/INST]'
srv          init: init: chat template, thinking = 1
main: model loaded
main: server is listening on http://127.0.0.1:8080
main: starting the main loop...
srv  update_slots: all slots are idle
srv  params_from_: Chat format: peg-native
slot get_availabl: id  3 | task -1 | selected slot by LRU, t_last = -1
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> temp-ext -> dist 
slot launch_slot_: id  3 | task 0 | processing task, is_child = 0
slot update_slots: id  3 | task 0 | new prompt, n_ctx_slot = 16384, n_keep = 0, task.n_tokens = 8186
slot update_slots: id  3 | task 0 | n_tokens = 0, memory_seq_rm [0, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 331, batch.n_tokens = 331, progress = 0.040435
srv  log_server_r: done request: POST /v1/chat/completions 127.0.0.1 200
slot update_slots: id  3 | task 0 | n_tokens = 331, memory_seq_rm [331, end)
srv  process_chun: processing audio...
encoding audio slice...
audio slice encoded in 767 ms
decoding audio batch 1/1, n_tokens_batch = 187
audio decoded (batch 1/1) in 1 ms
srv  process_chun: audio processed in 768 ms
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 518, batch.n_tokens = 0, progress = 0.063279
srv  update_slots: no tokens to decode
slot update_slots: id  3 | task 0 | n_tokens = 518, memory_seq_rm [518, end)
srv  process_chun: processing audio...
encoding audio slice...
audio slice encoded in 583 ms
decoding audio batch 1/1, n_tokens_batch = 187
audio decoded (batch 1/1) in 1 ms
srv  process_chun: audio processed in 584 ms
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 705, batch.n_tokens = 0, progress = 0.086123
srv  update_slots: no tokens to decode
slot update_slots: id  3 | task 0 | n_tokens = 705, memory_seq_rm [705, end)
srv  process_chun: processing audio...
encoding audio slice...
audio slice encoded in 589 ms
decoding audio batch 1/1, n_tokens_batch = 187
audio decoded (batch 1/1) in 1 ms
srv  process_chun: audio processed in 590 ms
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 892, batch.n_tokens = 0, progress = 0.108967
srv  update_slots: no tokens to decode
slot update_slots: id  3 | task 0 | n_tokens = 892, memory_seq_rm [892, end)
srv  process_chun: processing audio...
encoding audio slice...
audio slice encoded in 592 ms
decoding audio batch 1/1, n_tokens_batch = 187
audio decoded (batch 1/1) in 1 ms
srv  process_chun: audio processed in 593 ms
llama.cpp/tools/server/server-context.cpp:2686: fatal error - please provide logs and repro in https://github.com/ggml-org/llama.cpp/pull/20277

slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 1079, batch.n_tokens = 0, progress = 0.131810
srv  update_slots: no tokens to decode
(lldb) process attach --pid 59980
Process 59980 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00000001870b042c libsystem_kernel.dylib`__wait4 + 8
libsystem_kernel.dylib`__wait4:
->  0x1870b042c <+8>:  b.lo   0x1870b044c               ; <+40>
    0x1870b0430 <+12>: pacibsp 
    0x1870b0434 <+16>: stp    x29, x30, [sp, #-0x10]!
    0x1870b0438 <+20>: mov    x29, sp
Target 0: (llama-server) stopped.
Executable binary set to "llama.cpp/build/bin/llama-server".
Architecture set to: arm64-apple-macosx-.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00000001870b042c libsystem_kernel.dylib`__wait4 + 8
    frame #1: 0x0000000100d49558 libggml-base.0.9.8.dylib`ggml_abort + 156
    frame #2: 0x00000001003b9c98 llama-server`server_context_impl::update_slots() + 16072
    frame #3: 0x000000010037f864 llama-server`server_queue::start_loop(long long) + 504
    frame #4: 0x000000010030f788 llama-server`main + 14376
    frame #5: 0x0000000186d21d54 dyld`start + 7184
(lldb) quit
```
</details>

<!-- Long logs that you upload as files go here, outside the "console" area -->

--- a/tools/server/server-context.cpp
+++ b/tools/server/server-context.cpp
@@ -2493,7 +2493,7 @@ private:
                     bool has_mtmd = false;
 
                     // check if we should process the image
-                    if (slot.prompt.n_tokens() < slot.task->n_tokens() && input_tokens[slot.prompt.n_tokens()] == LLAMA_TOKEN_NULL) {
+                    while (slot.prompt.n_tokens() < slot.task->n_tokens() && input_tokens[slot.prompt.n_tokens()] == LLAMA_TOKEN_NULL) {
                         // process the image
                         size_t n_tokens_out = 0;
                         int32_t res = input_tokens.process_chunk(ctx, mctx, slot.prompt.n_tokens(), slot.prompt.tokens.pos_next(), slot.id, n_tokens_out);

